presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1202,9 +1205,30 @@ public PlanWithProperties visitUnion(UnionNode node, PreferredProperties parentP
                 // children partitioning and don't GATHER partitioned inputs
                 // TODO: add FIXED_ARBITRARY_DISTRIBUTION support on non empty singleNodeChildren
                 if (!parentPartitioningPreference.isPresent() || parentPartitioningPreference.get().isDistributed()) {
-                    return arbitraryDistributeUnion(node, distributedChildren, distributedOutputLayouts);
+                    // TODO: can we insert LOCAL exchange for one child SOURCE distributed and another HASH distributed?
+                    if (getNumberOfTableScans(distributedChildren) == 0 && isSameOrSystemCompatiblePartitions(extractRemoteExchangePartitioningHandles(distributedChildren))) {
+                        // No source distributed child, we can use insert LOCAL exchange
+                        // TODO: if all children have the same partitioning, pass this partitioning to the parent
+                        // instead of "arbitraryPartition".
+                        return new PlanWithProperties(node.replaceChildren(distributedChildren));
+                    }
+                    else if (preferDistributedUnion) {
+                        // Presto currently can not execute stage that has multiple table scans, so in that case
+                        // we have to insert REMOTE exchange with FIXED_ARBITRARY_DISTRIBUTION instead of local exchange
+                        return new PlanWithProperties(
+                                new ExchangeNode(
+                                        idAllocator.getNextId(),
+                                        REPARTITION,
+                                        REMOTE_STREAMING,
+                                        new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), node.getOutputVariables()),
+                                        distributedChildren,
+                                        distributedOutputLayouts,
+                                        false,
+                                        Optional.empty()));
+                    }
                 }
-
+                // TODO: We should support multiple table scans in a single fragment for efficient union implementation
```

Comment:
Maybe explain now there are two situations UNION will be executed by gathering into the same worker
1. Parent node doesn't prefer distributed partitioning (e.g. TableFinish)
2. Parent node prefers distributed partitioning  but `prefer_distributed_union` is disabled

For the second situation, we should support multiple table scans in a single fragment.